src/onyx/peer/function.clj

```diff
@@ -66,21 +62,66 @@
                 (count active-peers)))
       (rand-nth (operation/select-n-peers id active-peers max-downstream-links)))))
 
-;; Needs a performance boost
-(defmethod p-ext/write-batch :default
-  [{:keys [onyx.core/id onyx.core/results onyx.core/messenger onyx.core/job-id onyx.core/max-downstream-links] :as event}]
-  (let [leaves (fast-concat (map :leaves results))
-        egress-tasks (:egress-ids (:onyx.core/serialized-task event))]
-    (when-not (empty? leaves)
-      (let [replica @(:onyx.core/replica event)
-            segments (build-segments-to-send leaves)
-            groups (group-by (juxt :route :hash-group) segments)
-            allocations (get (:allocations replica) job-id)]
-        (doseq [[[route hash-group] segs] groups]
-          (let [peers (get allocations (get egress-tasks route))
-                active-peers (filter #(= (get-in replica [:peer-state %]) :active) peers)
-                target (pick-peer id active-peers hash-group max-downstream-links)]
-            (when target
-              (let [link (operation/peer-link event target)]
-                (onyx.extensions/send-messages messenger event link segs)))))
-        {}))))
+(defn read-batch [{:keys [onyx.core/messenger] :as event}]
+  {:onyx.core/batch (onyx.extensions/receive-messages messenger event)})
+
+(defn write-batch 
```

Comment:
See https://github.com/onyx-platform/onyx/blob/pext-protocol/src/onyx/plugin/core_async.clj#L38 https://github.com/onyx-platform/onyx/blob/pext-protocol/src/onyx/interop.clj#L7 and https://github.com/onyx-platform/onyx-benchmark/blob/master/src/java/onyx/plugin/RandomInputPlugin.java#L137. Potentially this should be done via an abstract class with an abstract method. I'm not sure if there's an easy way to create these in java though.

The only reason why we have two copies in there is because the version in Function doesn't require a lot of lookups because we can put a lot of the keys (replica, state, id, etc) in the Function record once and be done with it (i.e. it's an optimisation). I could move all the logic into one function and do the lookup outside though. I'll do that.